File: Problem9/erlang/special_pythagorean_triplet.erl
-module(special_pythagorean_triplet).
-include_lib("eunit/include/eunit.hrl").
-compile(export_all).

product_of_pythagorean_triplet_with_sum(N) ->
  MyList = pythagorean_triplets_below(N),
  SpecialTriplet = [X || X <- MyList, lists:sum(X) =:= N],
  multiply(lists:flatten(SpecialTriplet), 1).

pythagorean_triplets_below(N) ->
  [[X, Y, Z] || 
    X <- lists:seq(1,N-2), 
    Y <- lists:seq(X+1,N-1), 
    Z <- lists:seq(Y+1,N), 
    X + Y + Z =< N,
    X * X + Y * Y =:= Z * Z
  ].


multiply([], Acc) -> Acc;
multiply([H|T], Acc) ->
  multiply(T, H * Acc).

multiply_test() ->
  6 = multiply([1, 2, 3], 1).

File: Problem9/haskell/special_pythagorean_triplet.hs
product_of_special_pythagorean_triplet :: Integer -> Integer
product_of_special_pythagorean_triplet n = multiply(concatenate(pythagorean_triplets_below n), 1)

pythagorean_triplets_below :: Integer -> [[Integer]]
pythagorean_triplets_below n = [[x, y, z] | x <- [1..(n `div` 2)], y <- [x+1..(n `div` 2)], z <- [y+1..(n `div` 2)], x + y + z == n,  x^2 + y^2 == z^2]

multiply :: ([Integer], Integer) -> Integer
multiply([], acc) = acc
multiply(x:xs, acc) = multiply(xs, x * acc)

concatenate :: [[a]] -> [a]
concatenate([]) = []
concatenate(xss) = foldr (++) [] xss
